File: leetcode/algorithms/495_teemo_attacking/main.py
from typing import List


class TeemoAttacking:
    def findPoisonedDuration(self, timeSeries: List[int], duration: int) -> int:
        result = 0

        for i in range(1, len(timeSeries)):
            if timeSeries[i] - timeSeries[i - 1] >= duration:
                result += duration
            else:
                result += timeSeries[i] - timeSeries[i - 1]

        result += duration

        return result


    # Solution
    def solution(self, timeSeries: List[int], duration: int) -> int:
        ans = duration * len(timeSeries)

        for i in range(1, len(timeSeries)):
            ans -= max(0, duration - (timeSeries[i] - timeSeries[i - 1]))

        return ans

File: leetcode/algorithms/495_teemo_attacking/main.java
class TeemoAttacking {
    public int findPoisonedDuration(int[] timeSeries, int duration) {
        int result = 0;
        for (int i = 1; i < timeSeries.length; i++) {
            if (timeSeries[i] - timeSeries[i - 1] >= duration) {
                result += duration;
            } else {
                result += timeSeries[i] - timeSeries[i - 1];
            }
        }
        result += duration;
        return result;
    }


    // Best Solution
    public int bestSolution(int[] timeSeries, int duration) {
        if (timeSeries == null || timeSeries.length == 0 || duration == 0) return 0;

        int result = 0, start = timeSeries[0], end = timeSeries[0] + duration;
        for (int i = 1; i < timeSeries.length; i++) {
            if (timeSeries[i] > end) {
                result += end - start;
                start = timeSeries[i];
            }
            end = timeSeries[i] + duration;
        }
        result += end - start;

        return result;
    }
}